lib/evmone/eof.cpp

```diff
@@ -90,37 +151,261 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+                i += static_cast<size_t>(1 /* count */ + count * 2 /* tbl */);
+            }
+            else
+                return EOFValidationError::truncated_instruction;
+        }
+        else
+            i += instr::traits[op].immediate_size;
 
-    if (!instr::traits[op].is_terminating)
-        return EOFValidationError::missing_terminating_instruction;
+        if (i >= code.size())
+            return EOFValidationError::truncated_instruction;
+    }
 
     return EOFValidationError::success;
 }
 
+bool validate_rjump_destinations(bytes_view code) noexcept
+{
+    // Collect relative jump destinations and immediate locations
+    const auto code_size = code.size();
+    std::vector<size_t> rjumpdests;
+    std::vector<bool> immediate_map(code_size);
+
+    for (size_t i = 0; i < code_size; ++i)
+    {
+        const auto op = code[i];
+
+        if (op == OP_RJUMP || op == OP_RJUMPI)
+        {
+            const auto offset = read_int16_be(&code[i + 1]);
+            const auto jumpdest = static_cast<int32_t>(i) + 3 + offset;
+            if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+                return false;
+            rjumpdests.push_back(static_cast<size_t>(jumpdest));
+        }
+        else if (op == OP_RJUMPV)
+        {
+            constexpr auto REL_OFFSET_SIZE = sizeof(int16_t);
+
+            const auto count = code[i + 1];
+
+            const auto post_offset = 1 + 1 /* count */ + count * REL_OFFSET_SIZE /* tbl */;
+
+            for (size_t k = 0; k < count * REL_OFFSET_SIZE; k += REL_OFFSET_SIZE)
+            {
+                const auto rel_offset = read_int16_be(&code[i + 1 + 1 + static_cast<uint16_t>(k)]);
+                const auto jumpdest = static_cast<int>(i + post_offset) + rel_offset;
+                if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+                    return false;
+                rjumpdests.push_back(static_cast<size_t>(jumpdest));
+            }
+
+            const auto rjumpv_imm_size = size_t{1} + count * REL_OFFSET_SIZE;
+            std::fill_n(
+                immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, rjumpv_imm_size, true);
+            i += rjumpv_imm_size;
+            continue;
+        }
+
+        const auto imm_size = instr::traits[op].immediate_size;
+        std::fill_n(immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, imm_size, true);
+        i += imm_size;
+    }
+
+    // Check relative jump destinations against immediate locations.
+    for (const auto rjumpdest : rjumpdests)
+        if (immediate_map[rjumpdest])
+            return false;
+
+    return true;
+}
+
+std::pair<EOFValidationError, int32_t> validate_max_stack_height(
+    bytes_view code, size_t func_index, const std::vector<EOFCodeType>& funcs_in_outs)
+{
+    assert(!code.empty());
+
+    // Changing this type to smaller integer (uint16_t) have in mind that it can overflow.
```

Comment:
What does this comment mean?

Reply:
Initially it was uint16_t but after review it was decided to use bigger types. In general stack height must not be bigger than 2 bytes but int the validation we have to use bigger type

Reply:
```suggestion
    // Stack height in the header is limited to uint16_t, but keeping larger size for ease of calculation.
```

Reply:
done